---
title: "PHC 6937 (Data Visualization) Final Project"
author: "Jack Leary"
date: "`r Sys.Date()`"
output:
  html_document:
    theme: journal
    highlight: tango
    df_print: kable
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: true
    code_folding: show
    code_download: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      message = FALSE, 
                      warning = FALSE, 
                      fig.align = "center")
options(future.globals.maxSize = 2000 * 1024^2)
```

# Introduction

# Libaries

```{r}
library(dplyr)
library(Seurat)
library(ggpubr)
library(ggplot2)
library(scRNAseq)
library(SCISSORS)
library(patchwork)
library(jackknife)
library(paletteer)
library(kableExtra)
```

# Data

There are several samples here, so we'll need to create a dictionary of the samplenames and the sample types. 

```{r}
dir_names <- c("GSM4679532", "GSM4679533", "GSM4679534", "GSM4679535", "GSM4679536", 
               "GSM4679537", "GSM4679538", "GSM4679539", "GSM4679540", "GSM4679541", 
               "GSM4679542", "GSM4679543", "GSM4679544", "GSM4679545", "GSM4679546", "GSM4679547")
sample_names <- c("P01", "P02", "P03", "P04", "P05", "P06", "P07", "P08", "P09", "P10", 
                  "MET01", "MET02", "MET03", "MET04", "MET05", "MET06")
sample_types <- c(rep("Primary", 10), rep("Metastatic", 6))
sample_dict <- data.frame(Directory = dir_names, 
                          Sample = sample_names, 
                          Type = sample_types)
sample_dict %>% 
  kbl(booktabs = TRUE, caption = "Sample Dictionary") %>% 
  kable_minimal("hover", html_font = "Avenir", full_width = FALSE)
```

## Dataset Integration

First, we loop across the samples, create `Seurat` objects, and normalize expression and identify variable genes. The first 10 samples are primary tumors, and the next 6 are metastatic. 

```{r}
obj_list <- list()
for (i in seq(sample_dict$Directory)) {
  file <- paste0("~/Desktop/Data/Lin et al/GSE154778_RAW/", sample_dict$Directory[i])
  counts <- Read10X(file)
  seurat_obj <- CreateSeuratObject(counts, 
                                   project = sample_dict$Sample[i], 
                                   min.cells = 3, 
                                   min.features = 1000)
  seurat_obj@meta.data$sample <- sample_dict$Sample[i]
  seurat_obj[["percent_MT"]] <- PercentageFeatureSet(seurat_obj, pattern = "^MT-")
  seurat_obj <- SCTransform(seurat_obj, 
                            variable.features.n = 2000, 
                            vars.to.regress = "percent_MT", 
                            seed.use = 629, 
                            verbose = FALSE)
  obj_list[[i]] <- seurat_obj
}
primary_obj <- obj_list[1:10]
meta_obj <- obj_list[11:16]
```

```{r, echo=FALSE}
rm(counts, obj_list, seurat_obj, dir_names, file, sample_names, sample_types)
```

Next we integrate all the samples into one `Seurat` object. We'll start with the primary samples. 

```{r}
integration_genes <- SelectIntegrationFeatures(object.list = primary_obj, 
                                               nfeatures = 2000, 
                                               verbose = FALSE)
primary_obj <- PrepSCTIntegration(object.list = primary_obj, 
                                  anchor.features = integration_genes, 
                                  verbose = FALSE)
integration_anchors <- FindIntegrationAnchors(object.list = primary_obj, 
                                              normalization.method = "SCT", 
                                              anchor.features = integration_genes)
primary <- IntegrateData(anchorset = integration_anchors, 
                         normalization.method = "SCT", 
                         k.weight = 50,
                         verbose = FALSE)
DefaultAssay(primary) <- "integrated"
```

We repeat the process for the metastatic samples. 

```{r}
integration_genes <- SelectIntegrationFeatures(object.list = meta_obj, 
                                               nfeatures = 2000, 
                                               verbose = FALSE)
meta_obj <- PrepSCTIntegration(object.list = meta_obj, 
                               anchor.features = integration_genes, 
                               verbose = FALSE)
integration_anchors <- FindIntegrationAnchors(object.list = meta_obj, 
                                              normalization.method = "SCT", 
                                              anchor.features = integration_genes)
meta <- IntegrateData(anchorset = integration_anchors, 
                      normalization.method = "SCT", 
                      k.weight = 50,
                      verbose = FALSE)
DefaultAssay(meta) <- "integrated"
```

# Preprocessing

## Primary Tumors

We run PCA / UMAP, create the SNN graph representation of our cells, and cluster the cells using Louvain modularity optimization. 

```{r}
primary <- RunPCA(primary, 
                  npcs = 30, 
                  verbose = FALSE, 
                  seed.use = 629)
primary <- RunUMAP(primary, 
                   reduction = "pca", 
                   dims = 1:30, 
                   metric = "cosine",
                   min.dist = 0.1, 
                   n.components = 2, 
                   seed.use = 629, 
                   verbose = FALSE)
primary <- FindNeighbors(primary, 
                         reduction = "pca", 
                         dims = 1:30, 
                         annoy.metric = "cosine", 
                         verbose = FALSE)
primary <- FindClusters(primary, 
                        resolution = 0.3, 
                        random.seed = 629,
                        verbose = FALSE)
```

## Metastatic Tumors

Now we repeat the preprocessing on the metastatic tumors. 

```{r}
meta <- RunPCA(meta, 
               npcs = 30, 
               verbose = FALSE, 
               seed.use = 629)
meta <- RunUMAP(meta, 
                reduction = "pca", 
                dims = 1:30, 
                min.dist = 0.05, 
                metric = "cosine", 
                n.components = 2, 
                seed.use = 629, 
                verbose = FALSE)
meta <- FindNeighbors(meta, 
                      reduction = "pca", 
                      dims = 1:30, 
                      annoy.metric = "cosine", 
                      verbose = FALSE)
meta <- FindClusters(meta, 
                     resolution = 0.25, 
                     verbose = FALSE, 
                     random.seed = 629)
```

## Visualization

First we visualize the semi-supervised clustering we've generated using Louvain modularity optimization. 

```{r}
p0 <- DimPlot(primary) + 
      labs(title = "Primary Tumors", x = "UMAP 1", y = "UMAP 2") + 
      scale_color_paletteer_d("ggthemes::Classic_20") + 
      theme_yehlab() + 
      theme(text = element_text(family = "Avenir"), 
            axis.title.x = element_text(hjust = 0, size = 10), 
            axis.title.y = element_text(hjust = 0, size = 10), 
            plot.title = element_text(hjust = 0.5, size = 12)) + 
      guides(color = guide_legend(nrow = 2, byrow = TRUE, override.aes = list(size = 3)))
p1 <- DimPlot(meta) + 
      labs(title = "Metastatic Tumors") + 
      scale_color_paletteer_d("ggthemes::Classic_20") + 
      theme_yehlab() + 
      theme(axis.title = element_blank(), 
            text = element_text(family = "Avenir"), 
            plot.title = element_text(hjust = 0.5, size = 12)) + 
      guides(color = guide_legend(nrow = 2, byrow = TRUE, override.aes = list(size = 3)))
(p0 | p1) + plot_annotation(caption = "Louvain Clustering", 
                            theme = theme(plot.caption = element_text(face = "italic")))
```

Next we check out how the samples are arranged in UMAP space. We see that there is next to no grouping by sample, which tells us that the data integration process worked correctly and the sample-related batch effect has been regressed out. 

```{r}
p2 <- DimPlot(primary, group.by = "sample") + 
      labs(x = "UMAP 1", y = "UMAP 2", title = "Sample IDs (Primary Tumors)") + 
      scale_color_paletteer_d("miscpalettes::pastel") + 
      theme_yehlab() + 
      theme(text = element_text(family = "Avenir"), 
            axis.title = element_text(hjust = 0, size = 10), 
            plot.title = element_text(face = "bold", hjust = 0.5, size = 12), 
            legend.text = element_text(size = 10)) + 
      guides(color = guide_legend(nrow = 2, byrow = TRUE, override.aes = list(size = 3)))
p3 <- DimPlot(meta, group.by = "sample") + 
      labs(title = "Sample IDs (Metastatic Tumors)") + 
      scale_color_paletteer_d("miscpalettes::pastel") + 
      theme_yehlab() + 
      theme(axis.title = element_blank(), text = element_text(family = "Avenir"), 
            plot.title = element_text(face = "bold", hjust = 0.5, size = 12), 
            legend.text = element_text(size = 10)) + 
      guides(color = guide_legend(nrow = 2, byrow = TRUE, override.aes = list(size = 3)))
(p2 | p3) + plot_annotation(theme = theme(plot.caption = element_text(face = "italic")))
```

# Analysis 

## Differential Expression

### Primary Tumor 

We use a Wilcox test for differential expression, and filter the results after the Bonferroni *p*-value correction at the $q < 0.01$ level. 

```{r}
primary_markers <- FindAllMarkers(primary, 
                                  logfc.threshold = 3, 
                                  test.use = "wilcox", 
                                  only.pos = TRUE, 
                                  random.seed = 629, 
                                  verbose = FALSE) %>% 
                   filter(p_val_adj < 0.01)
primary_markers %>% 
  group_by(cluster) %>% 
  top_n(n = 5, wt = avg_log2FC) -> primary_markers_top5
p4 <- DotPlot(primary, 
              cols = c("grey70", "firebrick"), 
              features = primary_markers_top5$gene) + 
      labs(x = "Gene", 
           y = "Cluster", 
           color = "Mean Expression", 
           size = "Percent Expressed", 
           title = "Primary Tumor Marker Genes") + 
      theme(legend.position = "bottom", 
            axis.line = element_blank(), 
            legend.justification = "center", 
            text = element_text(family = "Avenir"), 
            axis.title.y = element_text(angle = 360, vjust = 0.5), 
            panel.border = element_rect(fill = NA, size = 1, color = "black"), 
            axis.text.x = element_text(angle = 45, size = 8, vjust = 1, hjust = 1)) + 
      guides(color = guide_colorbar(title.position = "top", title.hjust = 0.5, barwidth = unit(5, "cm")), 
             size = guide_legend(title.position = "top", title.hjust = 0.5))
```

```{r}
p4
```

### Metastatic Tumor

We perform the same DE test on the metastatic tumor cells. 

```{r}
meta_markers <- FindAllMarkers(meta, 
                               logfc.threshold = 3, 
                               test.use = "wilcox", 
                               only.pos = TRUE, 
                               random.seed = 629, 
                               verbose = FALSE) %>% 
                filter(p_val_adj < 0.01)
meta_markers %>% 
  group_by(cluster) %>% 
  top_n(n = 5, wt = avg_log2FC) -> meta_markers_top5
p5 <- DotPlot(meta, 
              cols = c("grey70", "firebrick"), 
              features = meta_markers_top5$gene) + 
      labs(x = "Gene", 
           y = "Cluster", 
           color = "Mean Expression", 
           size = "Percent Expressed", 
           title = "Metastatic Tumor Marker Genes") + 
      theme(legend.position = "bottom", 
            axis.line = element_blank(), 
            legend.justification = "center", 
            legend.title = element_text(size = 12), 
            text = element_text(family = "Avenir"), 
            plot.title = element_text(hjust = 0.5), 
            axis.title.y = element_text(angle = 360, vjust = 0.5), 
            panel.border = element_rect(fill = NA, size = 1, color = "black"), 
            axis.text.x = element_text(angle = 45, size = 8, vjust = 1, hjust = 1)) + 
      guides(color = guide_colorbar(title.position = "top", title.hjust = 0.5, barwidth = unit(5, "cm")), 
             size = guide_legend(title.position = "top", title.hjust = 0.5))
```

We can see TYROBP & C1QA, which are myeloid markers, in cluster 4. Cluster 5 shows CD3D & NKG7; this indicates that the cluster is composed of T/NK cells. Cluster 2 has TOP2A, which is a DC marker. 

```{r}
p5
```

## Cell Type Identification 

### SingleR 

We'll use the labeled reference data from Baron *et al* (2016) in order to give broad cell types 

#### Primary Tumor

```{r}
pdac_ref <- readRDS("/Volumes/labs/Home/Jen Jen Yeh Lab/Jack/scRNAseq/Seurat/single_cell_ref_normalized.Rds")
cell_preds <- SingleR(test = primary@assays$integrated@scale.data, 
                      ref = pdac_ref, 
                      labels = pdac_ref$label, 
                      clusters = primary$seurat_clusters, 
                      de.method = "wilcox")
primary[["SingleR"]] <- cell_preds$labels[match(primary[[]][["seurat_clusters"]], rownames(cell_preds))]
primary[["SingleR"]] <- case_when(primary[["SingleR"]] == "acinar" ~ "Acinar", 
                                  primary[["SingleR"]] == "activated_stellate" ~ "Activated Stellate", 
                                  primary[["SingleR"]] == "ductal" ~ "Ductal", 
                                  primary[["SingleR"]] == "endothelial" ~ "Endothelial", 
                                  primary[["SingleR"]] == "macrophage" ~ "Macrophage", 
                                  primary[["SingleR"]] == "mast" ~ "Mast", 
                                  primary[["SingleR"]] == "quiescent_stellate" ~ "Quiescent Stellate")
p6 <- DimPlot(primary, group.by = "SingleR") + 
      labs(x = "UMAP 1", y = "UMAP 2", subtitle = "Primary Tumor") + 
      scale_color_paletteer_d("ggthemes::Classic_20") + 
      theme_yehlab() + 
      theme(legend.position = "right", 
            plot.title = element_blank(), 
            legend.text = element_text(size = 10), 
            text = element_text(family = "Avenir"), 
            plot.subtitle = element_text(hjust = 0.5), 
            axis.title = element_text(hjust = 0, size = 9)) + 
      guides(color = guide_legend(ncol = 1, override.aes = list(size = 3)))
```

```{r}
p6
```

#### Metastatic Tumor

```{r}
cell_preds <- SingleR(test = meta@assays$integrated@scale.data, 
                      ref = pdac_ref, 
                      labels = pdac_ref$label, 
                      clusters = meta$seurat_clusters, 
                      de.method = "wilcox")
meta[["SingleR"]] <- cell_preds$labels[match(meta[[]][["seurat_clusters"]], rownames(cell_preds))]
meta[["SingleR"]] <- case_when(meta[["SingleR"]] == "epsilon" ~ "Epsilon", 
                               meta[["SingleR"]] == "t_cell" ~ "T", 
                               meta[["SingleR"]] == "ductal" ~ "Ductal", 
                               meta[["SingleR"]] == "macrophage" ~ "Macrophage")
p7 <- DimPlot(meta, group.by = "SingleR", cols = paletteer_d("ggthemes::Classic_20")[8:11]) + 
      labs(subtitle = "Metastatic Tumor") + 
      theme_yehlab() + 
      theme(legend.position = "right", 
            plot.title = element_blank(), 
            axis.title = element_blank(), 
            legend.text = element_text(size = 9), 
            text = element_text(family = "Avenir"), 
            plot.subtitle = element_text(hjust = 0.5)) + 
      guides(color = guide_legend(ncol = 1, override.aes = list(size = 3)))
```

```{r}
(p7 / p6) + plot_annotation(title = "SingleR Cell Labels", 
                            theme = theme(plot.title = element_text(family = "Avenir", size = 16, hjust = 0.5)))
```

## CNV Estimation 

